### PR TITLE
Track State Updates

### PR DESCRIPTION
**Pull request by hiroshihorie**
@davidzhao @boks1971 @davidliu 
What is the desired behavior for this signal ?

### CODE REVIEW

**Comment by boks1971** [comment]
@hiroshihorie I think we need to emit (would be good at both room level and participant level) signals about video tracks of a particular remote participant pausing and resuming (this is pause/resume due to bandwidth constraints in the downstream channel).

Interested in thoughts from the Davids :-)

**Comment by hiroshihorie** [comment]
Is this feature turn on by default by the server ?

**Comment by boks1971** [comment]
> Is this feature turn on by default by the server ?

Yes @hiroshihorie .

**Comment by davidliu** [comment]
@boks1971 So to be clear, the end goal of this protocol message simply just be to expose something for devs to know that the stream is currently paused, or is there anything else we would want to do with it? 

This sounds like it would be a separate flag from the `muted` flag that we already have on TrackPublications, am I correct?

**Comment by boks1971** [comment]
> @boks1971 So to be clear, the end goal of this protocol message simply just be to expose something for devs to know that the stream is currently paused, or is there anything else we would want to do with it?

Nope, the goal is to notify/emit events for devs to know so that they can give some treatment to the video element (maybe change frozen video to grayscale. - we used to do this in Slack calls a long time back).

In the future, there may be more in the SDK, but nothing planned for now. So, notification to devs is the only thing to do.

> 
> This sounds like it would be a separate flag from the `muted` flag that we already have on TrackPublications, am I correct?

Yes it is. `muted` is a client initiated thing. It could be
- Publisher muting which essentially becomes `mute` for all subscribers as there is no media track from source
- Subscriber muting which is client initiated.

The `paused` state in this update is purely server side decision. Currently, the only reason this is paused is due to bandwidth constraint. In the future, they may be other business rules or admin controls which may pause without a client initiating a mute.

So, in general, it is good to think of it as
- Client initiated mute/pause
- Server initiated pause
and for every situation ask the question "is it clear to devs why the stream is paused/stopped? i. e. is it clear that it is client side initiated vs server side initiated"

Please let me know if this is all confusing.

File: Sources/LiveKit/Core/Room.swift [hunk 1]
```diff
@@ -345,4 +345,22 @@ extension Room: EngineDelegate {
...
         notify { $0.room(self, didFailToConnect: error) }
     }
+
+    func engine(_ engine: Engine, didUpdate trackStates: [Livekit_StreamStateInfo]) {
+
+        for trackState in trackStates {
+            // Try to find participant
+            let participant: Participant?
+
+            if trackState.participantSid == localParticipant?.sid {
```
Comment (boks1971) [inline Sources/LiveKit/Core/Room.swift:355]: You will not get this on local participant.

**Review by boks1971 — APPROVED** [review]
Is there no auto generated code for proto changes in ios?

File: Sources/LiveKit/Core/Room.swift (hunk 1 above)
Comment (hiroshihorie) [reply Sources/LiveKit/Core/Room.swift:355]: Oh yes so this is fore Remote only

**Comment by hiroshihorie** [comment]
I'm just emitting event and updating `RemoteParticipant.streamState` property. Is that all for now ? 😁

**Comment by boks1971** [comment]
Yeah that's it. I thought @davidliu was also emitting event at room level. But my memory is vague although I reviewed that only yesterday😊

**Comment by hiroshihorie** [comment]
ok thanks onichan 😁🙏

**Pull request merged**